### PR TITLE
[graphql] refactor node def, config, and type resolution from assets

### PR DESCRIPTION
**Pull request by alangenfeld**
go through the context for resolving pieces of information to facilitate different backing implementations in other workspace classes.

## How I Tested These Changes

existing tests

### CODE REVIEW

**Comment by alangenfeld** [comment]
* **#25552** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25552?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @alangenfeld and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/dagster-graphql/dagster_graphql/schema/asset_graph.py [hunk 1]
```diff
@@ -391,36 +391,42 @@ def stale_status_loader(self) -> StaleStatusLoader:
     def asset_graph_differ(self) -> Optional[AssetGraphDiffer]:
         return self._asset_graph_differ
 
+    def _job_selector(self) -> Optional[JobSelector]:
+        if len(self._asset_node_snap.job_names) < 1:
```
Comment (gibsondan) [inline python_modules/dagster-graphql/dagster_graphql/schema/asset_graph.py:395]: == 0?

File: python_modules/dagster-graphql/dagster_graphql/schema/config_types.py [hunk 2]
```diff
@@ -19,28 +19,26 @@
 
 
 def to_config_type(
-    config_schema_snapshot: ConfigSchemaSnapshot, config_type_key: str
+    get_config_type: Callable[[str], ConfigTypeSnap],
```
Comment (gibsondan) [inline python_modules/dagster-graphql/dagster_graphql/schema/config_types.py:22]: maybe worth a comment about why this is deferred?

**Review by gibsondan — APPROVED** [review]
Gnarly but impressive. I think we may still also want to defer the definitions fields that are using these types? It seems easy to accidentally slip in a new callsite to the node

File: python_modules/dagster-graphql/dagster_graphql/schema/pipelines/resource.py [hunk 3]
```diff
@@ -14,27 +17,23 @@ class GrapheneResource(graphene.ObjectType):
...
         self.name = resource_def_snap.name
         self.description = resource_def_snap.description
 
     def resolve_configField(self, _graphene_info: ResolveInfo):
-        if (
-            self._resource_def_snap.config_field_snap
-            # config type may not be present if mode config mapped
-            and self._config_schema_snapshot.has_config_snap(
-                self._resource_def_snap.config_field_snap.type_key
-            )
```
Comment (alangenfeld) [inline python_modules/dagster-graphql/dagster_graphql/schema/pipelines/resource.py:34]: [X]

**Pull request merged**